source/locationHelper.py

```diff
@@ -0,0 +1,437 @@
+
+	def toScreen(self, hwnd):
+		"""Converts self from client to screen coordinates and returns a new L{Point}"""
+		return Point(*winUser.ClientToScreen(hwnd, *self))
+
+class _RectMixin:
+	"""Mix-in class for properties shared between RectLTRB and RectLTWH classes"""
+
+	def toRECT(self):
+		"""Converts self to a L{ctypes.wintypes.RECT}"""
+		return RECT(self.left,self.top,self.right,self.bottom)
+
+	def toLogical(self, hwnd):
+		left,top=self.topLeft.toLogical(hwnd)
+		right,bottom=self.bottomRight.toLogical(hwnd)
+		if isinstance(self, RectLTWH):
+			return RectLTWH(left,top,right-left,bottom-top)
+		return RectLTRB(left,top,right,bottom)
+
+	def toPhysical(self, hwnd):
+		left,top=self.topLeft.toPhysical(hwnd)
+		right,bottom=self.bottomRight.toPhysical(hwnd)
+		if isinstance(self, RectLTWH):
+			return RectLTWH(left,top,right-left,bottom-top)
+		return RectLTRB(left,top,right,bottom)
+
+	def toClient(self, hwnd):
+		left, top =self.topLeft.toClient(hwnd)
+		if isinstance(self, RectLTWH):
+			return RectLTWH(left, top, self.width, self.height)
+		return RectLTRB(left, top, left+self.width, top+self.height)
+
+	def toScreen(self, hwnd):
+		left,top=self.topLeft.toScreen(hwnd)
+		if isinstance(self, RectLTWH):
+			return RectLTWH(left, top, self.width, self.height)
+		return RectLTRB(left, top, left+self.width, top+self.height)
+
+	@property
+	def topLeft(self):
+		return Point(self.left,self.top)
+
+	@property
+	def topRight(self):
+		return Point(self.right,self.top)
+
+	@property
+	def bottomLeft(self):
+		return Point(self.left,self.bottom)
+
+	@property
+	def bottomRight(self):
+		return Point(self.right,self.bottom)
+
+	@property
+	def center(self):
+		return Point(int(round(self.left+self.width/2.0)), int(round(self.top+self.height/2.0)))
+
+	def __contains__(self,other):
+		"""Returns whether other is a part of this rectangle."""
+		if isinstance(other,POINT_CLASSES):
+			return self.left <= other.x < self.right and  self.top <= other.y < self.bottom
+		if not isinstance(other,RECT_CLASSES):
+			try:
+				other=toRectLTRB(other)
+			except:
+				return False
+		return self.isSuperset(other) and self!=other
+
+	def isSubset(self,other):
+		"""Returns whether this rectangle is a subset of other (i.e. whether all points in this rectangle are contained by other)."""
+		if not isinstance(other,RECT_CLASSES):
+			try:
+				other=toRectLTRB(other)
+			except ValueError:
+				return False
+		return self.left >= other.left and self.top >= other.top and self.right <= other.right and self.bottom <= other.bottom
+
+	def isSuperset(self,other):
+		"""Returns whether this rectangle is a superset of other (i.e. whether all points of other are contained by this rectangle)."""
+		if not isinstance(other,RECT_CLASSES):
+			try:
+				other=toRectLTRB(other)
+			except ValueError:
+				return False
+		return other.left >= self.left and other.top >= self.top and other.right <= self.right and other.bottom <= self.bottom

```

Comment:
I believe there is a bug in here (both superset and subset ), caused by a mix up of top and bottom comparisons.

Consider:
```
r1=RectLTRB(left=10, top=14, right=14, bottom=10)
r2= RectLTRB(left=11, top=15, right=13, bottom=9)
```

- `r1.isSubset(r2)` returns `False`
- `r1.isSuperset(r2)` returns `True`
- `r2 in r1` returns `True`

And:
```
r3 = RectLTRB(left=10, top=14, right=14, bottom=10)
r4 = RectLTRB(left=11, top=13, right=13, bottom=11)
```
then:
- `r3.isSubset(r4)` returns `False`
- `r3.isSuperset(r4)` returns `False`
- `r4 in r3` returns `False`

Fix:
For the top and bottom comparisons, `other` and `self` should be swapped to the other side of the >= operator. Notice that the `__contains__` code for Point has this correct.

Reply:
> For the top and bottom comparisons, other and self should be swapped to the other side of the >= operator. Notice that the __contains__ code for Point has this correct.

The first question I have is, are these rectangles, where the top is lower than the bottom, considered to be valid? If not, I'd go for something like this

Furthermore, could you elaborate on the swapping of self and other? I'm reading this as you mean it to be 
```
return other.left >= self.left and self.top >= other.top and other.right <= self.right and self.bottom <= other.bottom
```
But that doesn't seem to be valid

How about:

```
return self.left<=other.left<=other.right<=self.right and self.top<=other.top<=other.bottom<=self.bottom
``

Reply:
> are these rectangles, where the top is lower than the bottom, considered to be valid?
Whether the top is lower than the bottom is matter of perspective, where `0,0` is the top left and y values decreasing vertically, then yes a top of 14 is below a bottom of 10. However `0,0` in the bottom left  and y values increasing vertically, then 14 is above 10.

I hadn't really considered the coordinate space this was supposed to be working in when considering this bug. If I swap the top and bottom values for those rectangles (r1 and r2) the results are as expected.

One way to make the subset and superset functions more robust would be to calculate based on each corner being contained by the rectangle:
`return all(p in self for p in [other.topLeft, other.topRight, other.bottomLeft, other.bottomRight])`

But I think it might be wise to include the coordinate space with the object. When a rect is created from some other object it can be set automatically, however when it is created by hand, or an unsupported type, then the developer needs to consider the address space.

Reply:
>But I think it might be wise to include the coordinate space with the object.

You mean as a property, and if so, what spaces should we support?

I've been thinking about creating overrides for __new__ in order to allow providing additional parameters, so we can also support setting a property that indicates whether provided coordinates are screen/client and physical/logical.

Reply:
I think there are really just two likely coordinate systems that will be used:
- Top left origin, y increases down the screen.
- Bottom left origin, y increases up the screen

I was thinking something like:
```
# when converting from some other library type, the space does not need to be mentioned
r1 = toRectLTWH(someScreenSpaceType)
assert(r1.coordSpace == TOP_LEFT_ORIGIN)
# when creating manually, the space is mentioned and disambiguates "top" and "bottom", and forces the developer to consider it.
r2 = RectLTWH(left=10, top=13, width=5, height=3, coordSpace=BOTTOM_LEFT_ORIGIN)
assert(r2.bottom == 10)
assert(r2.right == 15)
# then same as r2 but other coord system
r3 = RectLTWH(left=10, top=13, width=5, height=3, coordSpace=BOTTOM_LEFT_ORIGIN)
assert(r3.bottom == 16)
assert(r3.right == 15)
# Even though the values for bottom do not match, they refer to the same space
assert(r2.bottom == r3.bottom) # this is actually a problem. This really depends on the size of the space, whether 10 from the bottom is equal to 16 from the top is really dependent on the size of the screen... maybe things like:
r2.subset(r3)  # should just result in an exception?
```

Reply:
I'm honestly not convinced that we need a system for rectangles based on bottom left origin. Do you have reasons to think that they are currently part of NVDA core or will be in future?

Reply:
I dont know that we do... but I also dont know that we dont. It's the ambiguity that worries me. I would be fine with saying that this library does not support them, but considering how easy it is to miss a docstring, and make all kinds of mistakes, I would like to see if we can find a way to make it very clear to users of this library how they should be thinking about the values.

Reply:
How about raising a ValueError when creating a rectangle with right<left and bottom<top for RectLTRB? I'm afraid it is much harder for RectLTWH to enforce clarity, though.

Actually, I have seen an object location of 0, 0, -32000, -32000 in the wild, but these locations don't make sense to me. It looked a bit like giving a location for something we do not know the location of.

Reply:
A ValueError would be enough for `RectLTRB`. For the `RectLTWH` I can't think of anyway to check as it is currently, only with the introduction of an extra construction parameter / or a name change.